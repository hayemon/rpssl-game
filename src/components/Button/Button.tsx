import React, { useRef, useState } from 'react'
import { useDispatch } from 'react-redux'
import { resetResults, setLocked, setPlayerChoice } from 'store/actions'
import { usePlayMutation } from 'store/slices/apiSlice'
import { selectLocked } from 'store/slices/gameSlice'
import Choice from 'types/Choice'
import urlMapper from 'utils/urlMapper'
import styles from './index.module.scss'

type Props = {
  choice: Choice
}

const Button: React.FC<Props> = ({ choice }) => {
  const buttonRef = useRef<HTMLButtonElement>(null)

  const dispatch = useDispatch()
  const [play] = usePlayMutation()
  const locked = selectLocked()
  const [disabled, setDisabled] = useState<boolean>(false)

  const handleClick = () => {
    if (locked) {
      return
    }
    dispatch(setPlayerChoice(undefined))
    dispatch(resetResults())
    dispatch(setLocked(true))

    setTimeout(() => {
      dispatch(setPlayerChoice(choice.name))
    }, 100)

    setTimeout(() => {
      buttonRef.current?.blur()
      setDisabled(true)
    }, 600)

    setTimeout(() => {
      play(choice.id)
      dispatch(setLocked(false))
      setDisabled(false)
    }, 3000)
  }

  return (
    <button
      role='button'
      ref={buttonRef}
      className={`glass ${styles.box}`}
      disabled={disabled}
      onClick={handleClick}
    >
      <img src={urlMapper[choice.name]} />
    </button>
  )
}

export default Button
