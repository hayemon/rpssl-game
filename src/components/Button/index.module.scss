@import '../../mixins';

.box {
  width: 100px;
  height: 100px;
  border-radius: $border-radius;
  margin: $margin;
  padding: $padding;
  display: flex;
  justify-content: center;
  align-items: center;
  transition: 0.2s;
  overflow: hidden;
}

.box:focus {
  animation: slide-up 0.5s linear;
}

@keyframes slide-up {
  10% {
    transform: translateX(5px);
  }
  20% {
    transform: translateX(-5px) translateY(5px);
  }
  30% {
    transform: translateX(5px) translateY(-5px);
  }
  40% {
    transform: translateX(0) translateY(20px) scale(1.2);
  }
  60% {
    transform: translateY(0) scale(1);
    background-color: white;
  }
  80% {
    transform: translateY(-40vh);
    opacity: 0;
    background-color: $accent-color;
  }
  100% {
    transform: translateY(-200vh);
    opacity: 0;
  }
}

.box:not(:focus):hover::before {
  @include spinner-before;
}

.box:not(:focus):hover::after {
  @include spinner-after;
}

.box img {
  @include img-in-box;
}
