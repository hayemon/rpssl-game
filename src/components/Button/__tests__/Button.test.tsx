import '@testing-library/jest-dom/extend-expect'
import { fireEvent, waitFor } from '@testing-library/react'
import { renderWithProviders } from 'store/test/testUtils'
import Button from '../Button'

describe('Button', () => {
  it('renders', () => {
    const { asFragment } = renderWithProviders(
      <Button choice={{ id: 1, name: 'rock' }} />
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it('click button', async () => {
    const { getAllByRole, store } = renderWithProviders(
      <Button choice={{ id: 1, name: 'rock' }} />
    )

    const buttons = getAllByRole('button')

    fireEvent.click(buttons[0])

    const state = store.getState()
    expect(state.game.playerChoice).toBe(undefined)
    expect(state.game.results).toBe(undefined)
    expect(state.game.locked).toBe(true)

    await waitFor(() => {
      const state = store.getState()
      expect(state.game.playerChoice).toBe('rock')
      expect(state.game.locked).toBe(true)
    })
  })
})
