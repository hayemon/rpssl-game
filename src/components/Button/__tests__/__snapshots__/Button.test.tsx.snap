// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders 1`] = `
<DocumentFragment>
  <button
    class="glass undefined"
    role="button"
  >
    <img
      src="test-file-stub"
    />
  </button>
</DocumentFragment>
`;
